Cleaned and fixed history for new .replace rule implementation

valdisvi opened the pull request.

rhdunn commented on src/libespeak-ng/translate.c:
```diff
@@ -1788,78 +1788,159 @@ static int EmbeddedCommand(unsigned int *source_index_out)
+				curchar += used;
+				continue;
+			}
+			if ((upper_case = iswupper(curint)) == true) { // remember, if uppercase
```
[Minor] A single-statement if statement does not need braces. This is to keep a consistent coding style with the rest of the codebase.

rhdunn commented on src/libespeak-ng/translate.c:
```diff
@@ -1788,78 +1788,159 @@ static int EmbeddedCommand(unsigned int *source_index_out)
-			}
-		}
+	// if requested, print trace
+	if (option_phonemes & espeakPHONEMES_TRACE) {
```
[Minor] A single-statement if statement does not need braces. This is to keep a consistent coding style with the rest of the codebase.

rhdunn commented on src/libespeak-ng/translate.c:
```diff
@@ -1788,78 +1788,159 @@ static int EmbeddedCommand(unsigned int *source_index_out)
+	char charfrom[N_WORD_BYTES + 1] = { };
+	int usedfrom = 0;
+	int *pintfrom = from;
+	char *pcharfrom = &charfrom;
```
`&` is not needed when casting `char []` to `char *` (gcc 7.3.0 warning):
```
src/libespeak-ng/translate.c:1839:20: warning: initialization from incompatible pointer type [-Wincompatible-pointer-types]
  char *pcharfrom = &charfrom;
                    ^
```

rhdunn commented on src/libespeak-ng/translate.c:
```diff
@@ -1788,78 +1788,159 @@ static int EmbeddedCommand(unsigned int *source_index_out)
+	char charto[N_WORD_BYTES + 1] = { };
+	int usedto = 0;
+	int *pintto = to;
+	char *pcharto = &charto;
```
`&` is not needed when casting `char []` to `char *` (gcc 7.3.0 warning):
```
src/libespeak-ng/translate.c:1849:18: warning: initialization from incompatible pointer type [-Wincompatible-pointer-types]
  char *pcharto = &charto;
                  ^
```

rhdunn commented on src/libespeak-ng/translate.c:
```diff
@@ -1788,78 +1788,159 @@ static int EmbeddedCommand(unsigned int *source_index_out)
+
+	// Set up replacement in temporary buffer
+	char *psource = *matchstart;
+	char *ptmpsource = &tmpsource;
```
`&` is not needed when casting `char []` to `char *` (gcc 7.3.0 warning):
```
src/libespeak-ng/translate.c:1857:21: warning: initialization from incompatible pointer type [-Wincompatible-pointer-types]
  char *ptmpsource = &tmpsource;
                     ^
```

rhdunn commented on src/libespeak-ng/translate.c:
```diff
@@ -1788,78 +1788,159 @@ static int EmbeddedCommand(unsigned int *source_index_out)
+	// Set up replacement in temporary buffer
+	char *psource = *matchstart;
+	char *ptmpsource = &tmpsource;
+	pcharto = &charto;
```
`&` is not needed when casting `char []` to `char *` (gcc 7.3.0 warning):
```
src/libespeak-ng/translate.c:1858:10: warning: assignment from incompatible pointer type [-Wincompatible-pointer-types]
  pcharto = &charto;
          ^
```

rhdunn commented on src/libespeak-ng/translate.c:
```diff
@@ -1788,78 +1788,159 @@ static int EmbeddedCommand(unsigned int *source_index_out)
-		upper_case = 1;
+	// Write temporary buffer back to source
+	psource = *matchstart;
+	ptmpsource = &tmpsource;
```
`&` is not needed when casting `char []` to `char *` (gcc 7.3.0 warning):
```
src/libespeak-ng/translate.c:1883:13: warning: assignment from incompatible pointer type [-Wincompatible-pointer-types]
  ptmpsource = &tmpsource;
             ^
```

rhdunn commented on src/libespeak-ng/translate.c:
```diff
@@ -1788,78 +1788,159 @@ static int EmbeddedCommand(unsigned int *source_index_out)
+	char *psource = *matchstart;
+	char *ptmpsource = &tmpsource;
+	pcharto = &charto;
+	int replace = 0;
```
This variable is not used. Remove it?
```
src/libespeak-ng/translate.c:1859:6: warning: unused variable ‘replace’ [-Wunused-variable]
  int replace = 0;
      ^~~~~~~
```

rhdunn commented on src/libespeak-ng/translate.c:
```diff
@@ -1788,78 +1788,159 @@ static int EmbeddedCommand(unsigned int *source_index_out)
-	if (ignore_next) {
-		ignore_next = false;
-		return 8;
+static void ReplaceMatchingChars(char **matchstart, int *from, int *to, int upper_case)
```
`upper_case` is not used. If it is not required, remove it.
```
src/libespeak-ng/translate.c:1826:77: warning: unused parameter ‘upper_case’ [-Wunused-parameter]
 static void ReplaceMatchingChars(char **matchstart, int *from, int *to, int upper_case)
                                                                             ^~~~~~~~~~
```

rhdunn commented on src/libespeak-ng/translate.c:
```diff
@@ -1788,78 +1788,159 @@ static int EmbeddedCommand(unsigned int *source_index_out)
+static void TranslateChars(Translator *tr, char *source)
+{
+	// Replace characters using Translator.replace_chars in passed source buffer
+	int *replace_chars = tr->langopts.replace_chars;
```
`replace_chars` is not used, so should be removed.
```
src/libespeak-ng/translate.c:1900:23: warning: initialization discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]
  int *replace_chars = tr->langopts.replace_chars;
                       ^~
src/libespeak-ng/translate.c:1900:7: warning: unused variable ‘replace_chars’ [-Wunused-variable]
  int *replace_chars = tr->langopts.replace_chars;
       ^~~~~~~~~~~~~
```

rhdunn commented on src/libespeak-ng/translate.c:
```diff
@@ -1788,78 +1788,159 @@ static int EmbeddedCommand(unsigned int *source_index_out)
+	int *to;
+	bool upper_case = false;
+
+	if ((from = tr->langopts.replace_chars) == NULL)
```
```
src/libespeak-ng/translate.c:1907:12: warning: assignment discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]
  if ((from = tr->langopts.replace_chars) == NULL)
            ^
```

rhdunn commented on src/libespeak-ng/translate.c:
```diff
@@ -1788,78 +1788,159 @@ static int EmbeddedCommand(unsigned int *source_index_out)
+{
+	// Replace characters using Translator.replace_chars in passed source buffer
+	int *replace_chars = tr->langopts.replace_chars;
+	char *mysource = source;
```
This variable is not used. Remove it?
```
src/libespeak-ng/translate.c:1901:8: warning: unused variable ‘mysource’ [-Wunused-variable]
  char *mysource = source;
        ^~~~~~~~
```

rhdunn commented on src/libespeak-ng/translate.c:
```diff
@@ -1788,78 +1788,159 @@ static int EmbeddedCommand(unsigned int *source_index_out)
-	return new_c;
 }
 
 static int TranslateChar(Translator *tr, char *ptr, int prev_in, unsigned int c, unsigned int next_in, int *insert, int *wordflags)
```
`wordflags` is now no longer used (after removing the `SubstituteChar` call). Therefore, it should be removed.
```
src/libespeak-ng/translate.c:1948:122: warning: unused parameter ‘wordflags’ [-Wunused-parameter]
  int TranslateChar(Translator *tr, char *ptr, int prev_in, unsigned int c, unsigned int next_in, int *insert, int *wordflags)
                                                                                                                    ^~~~~~~~~
```

rhdunn commented on src/libespeak-ng/translate.c:
```diff
@@ -2102,6 +2182,10 @@ void TranslateClause(Translator *tr, int *tone_out, char **voice_change)
 	words[0].length = k;
 
+	// handle .replace rule of xx_rules file
+	TranslateChars(tr, &source);
```
`&` is not needed when casting `char []` to `char *` (gcc 7.3.0 warning):
```
src/libespeak-ng/translate.c:2188:21: warning: passing argument 2 of ‘TranslateChars’ from incompatible pointer type [-Wincompatible-pointer-types]
  TranslateChars(tr, &source);
                     ^
src/libespeak-ng/translate.c:1897:13: note: expected ‘char *’ but argument is of type ‘char (*)[840]’
 static void TranslateChars(Translator *tr, char *source)
             ^~~~~~~~~~~~~~
```

rhdunn commented on src/libespeak-ng/translate.c:
```diff
@@ -1788,78 +1788,159 @@ static int EmbeddedCommand(unsigned int *source_index_out)
+			idx++;
+		//idx++;
+	}
+	*from = idx;
```
This logic (specifically the bit in the `next_step` block) is confusing to read. It would be easier to read if `*from` was used directly, instead of using an `idx` variable.

rhdunn commented on src/libespeak-ng/translate.c:
```diff
@@ -1788,78 +1788,159 @@ static int EmbeddedCommand(unsigned int *source_index_out)
+	*to = idx;
+	if (**to == 0)
+		return false;
+	return true;
```
[Minor] These 3 lines can be replaced by a single `return **to != 0;`.

rhdunn commented on src/libespeak-ng/translate.c:
```diff
@@ -1788,78 +1788,159 @@ static int EmbeddedCommand(unsigned int *source_index_out)
+	while (*idx == ' ') // skip remaining spaces, if exists
+		idx++;
+	//*to = ++idx;
+	*to = idx;
```
This is similarly confusing to the `from` logic for the same reason (it obscures where/how `to` is being modified).

It would be better if `*to = *from` was used before the first while loop, then use `*to` instead of `idx`.

valdisvi commented:
Reece, I fixed code issues you mentioned in review, and made these commits on top of latest master tree, to  fix merging issues and make history cleaner.
I couldn't fix warning 
```
src/libespeak-ng/translate.c:1901:12: warning: pointer targets in assignment differ in signedness [-Wpointer-sign]
  if ((from = tr->langopts.replace_chars) == NULL)
```
because in other cases utf-16 `int` is used as signed, but `tr->langopts.replace_chars` returns unsigned and explicit casting made just different warning.
Regarding to `upper_case` — I forgot to actually use it and fixed that, but, during this I found, that actually it cannot be used similarly as before. As now replacement can go from one letter to many before reading word, handling uppercase letters should be more precise (e.g. should check, if entire word is in uppercase). But current implementation will be correct for most of cases (i.e. word starts with capital letter, or, one to one letter is replaced) and I added `TODO` as a reminder.

rhdunn commented:
Merged, thanks.

Longer term, the `char8_t` (UTF-8), `char16_t` (UTF-16), and `char32_t` types should be used to make it clearer what variables are operating on unicode strings.
